ReactWindows/ReactNative/UIManager/UIViewOperationQueueCollection.cs

```diff
@@ -0,0 +1,564 @@
+/// </summary>
+
+using Newtonsoft.Json.Linq;
+using ReactNative.Bridge;
+using ReactNative.Bridge.Queue;
+using ReactNative.Modules.Core;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Windows.ApplicationModel.Core;
+using Windows.UI.Core;
+
+namespace ReactNative.UIManager
+{
+    /// <summary>
+    /// This class acts as a buffer for command executed on
+    /// <see cref="NativeViewHierarchyManager"/>. It exposes similar methods as
+    /// mentioned classes but instead of executing commands immediately, it 
+    /// enqueues those operations in a queue that is then flushed from 
+    /// <see cref="UIManagerModule"/> once a JavaScript batch of UI operations
+    /// is finished.
+    /// There is one queue instance created for each top level window since
+    /// each of these windows run on a different thread.
+    /// </summary>
+    public class UIViewOperationQueue
+    {
+        private readonly ReactContext _reactContext;
+        private readonly ViewManagerRegistry _viewManagerRegistry;
+
+        private class QueueInstanceInfo
+        {
+            public UIViewOperationQueueInstance queueInstance;
+            public int rootViewCount;
+        }
+
+        private UIViewOperationQueueInstance _mainUiViewOperationsQueueInstance;
+
+        // Maps tags to corresponding UIViewOperationQueueInstance
+        private readonly IDictionary<int, UIViewOperationQueueInstance> _reactTagToOperationQueue = new Dictionary<int, UIViewOperationQueueInstance>();
+
+        // Maps CoreDispatcher to corresponding UIViewOperationQueueInstance + rootView accounting
+        private readonly IDictionary<CoreDispatcher, QueueInstanceInfo> _dispatcherToOperationQueueInfo = new Dictionary<CoreDispatcher, QueueInstanceInfo>();
+
+        // Action queue used by UIManagerModule
+        private IActionQueue _uiManagerActionQueue;
+
+        private bool _active;
+
+        /// <summary>
+        /// Instantiates the <see cref="UIViewOperationQueue"/>.
+        /// </summary>
+        /// <param name="reactContext">The React context.</param>
+        /// The native view hierarchy manager.
+        /// </param>
+        public UIViewOperationQueue(ReactContext reactContext, ViewManagerRegistry viewManagerRegistry)
+        {
+            _reactContext = reactContext;
+            _viewManagerRegistry = viewManagerRegistry;
+            _active = false;
+
+            // Corner case: UWP scenarios that start with no main window.
+            // We create the UIViewOperationQueueInstance for main dispatcher thread ahead of time so animations
+            // in secondary windows can work.
+            var queueInfo = new QueueInstanceInfo()
+            {
+                queueInstance = new UIViewOperationQueueInstance(
+                    _reactContext,
+                    new NativeViewHierarchyManager(_viewManagerRegistry, DispatcherHelpers.MainDispatcher, OnViewsDropped),
```

Comment:
>OnViewsDropped [](start = 107, length = 14)

Should we send null for `OnViewsDropped` on the `MainDispatcher`? Presumably if we clean up the main window view operation queue, everything else should go with it.

Reply:
Just thinking that could allow us to implement this feature without having an impact on GC for single window apps.

---
In reply to: [167536716](https://github.com/Microsoft/react-native-windows/pull/1615#discussion_r167536716) [](ancestors = 167536716)

Reply:
I hear you.
Problem is JS land likes to delete subtrees, relying on native to do the recursion, etc.
So even in single window situations we have a growing map if not cleaned properly, in the current solution.

One more radical approach is to use the map just for secondary windows. If a tag is not in the map, it has to be in main.. This way you won't need a cleanup for main.
It's not that balanced behavior anymore, but it optimizes the main window. 

Should I go with that? The change itself is easy.

---
In reply to: [167536871](https://github.com/Microsoft/react-native-windows/pull/1615#discussion_r167536871) [](ancestors = 167536871,167536716)